File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.py
import sys
import math
sys.setrecursionlimit(10**6)

MAX_NUM = 1_000_000_001
def init(T,A,s,e,i):
    # T: Tree , a : Array , s : start, e : end, i : idx
    if s==e:
        T[i-1] = [A[s],s]
        return T[i-1]
    mid = (s+e)//2
    left = init(T,A,s,mid,i*2)
    right = init(T,A,mid+1,e,i*2+1)

    if left[0]<right[0]:
        T[i-1] = left
    else:
        T[i-1] = right
    return T[i-1]

def query(T,ns,ne,i,qs,qe):
    # ns : node start, ne : node end
    # qs : query stary, qe : query end
    if qs>ne or qe<ns:
        return [MAX_NUM,-1]
    if qs<=ns and qe>=ne:
        return T[i-1]
    mid = (ns+ne)//2
    left = query(T,ns,mid,i*2,qs,qe)
    right = query(T,mid+1,ne,i*2+1,qs,qe)
    if left[0] < right[0]:
        return left
    else:
        return right

def solve(s,e):
    minh,idx = query(tree,0,len(h)-1,1,s,e)
    a1,a2,a3 = (e-s+1)*minh,0,0
    if idx-1 >=s:
        a2 = solve(s,idx-1)
    if idx+1 <=e:
        a3 = solve(idx+1,e)

    return max(a1,a2,a3)

while True:
    h = list(map(int,sys.stdin.readline().split()))
    if len(h) == 1 and h[0] ==0:
        break
    n = h[0]
    h = h[1:]
    tree = [MAX_NUM]*(2**(math.ceil(math.log(n,2))+1))
    init(tree,h,0,n-1,1)
    ans = solve(0,n-1)
    print(ans)

File: 백준/Platinum/6549. 히스토그램에서 가장 큰 직사각형/히스토그램에서 가장 큰 직사각형.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    static int N;
    static int[] arr;
    static int[] tree;

    static int calc(int left, int right) {
        if (left == -1) {
            return right;
        } else if (right == -1) {
            return left;
        }

        if (arr[left] < arr[right]) {
            return left;
        }
        return right;
    }

    static int init(int idx, int s, int e) {
        if (s == e) {
            return tree[idx] = s;
        }
        int m = (s + e) / 2;
        int left = init(idx * 2, s, m);
        int right = init(idx * 2 + 1, m + 1, e);
        return tree[idx] = calc(left, right);
    }

    static int query(int ql, int qr, int idx, int tl, int tr) {
        if (qr < tl || tr < ql) {
            return -1;
        }
        if (ql <= tl && tr <= qr) {
            return tree[idx];
        }

        int m = (tl + tr) / 2;
        int left = query(ql, qr, idx * 2, tl, m);
        int right = query(ql, qr, idx * 2 + 1, m + 1, tr);
        return calc(left, right);
    }


    static long getMaxSquare(int s, int e) {
        if (s == e) {
            return arr[s];
        }
        if (s < e) {
            int minIdx = query(s, e, 1, 0, N - 1);
            long left = getMaxSquare(s, minIdx - 1);
            long right = getMaxSquare(minIdx + 1, e);

            long minArea = (long) arr[minIdx] * (e - s + 1);
            long maxValue = Math.max(left, right);
            maxValue = Math.max(maxValue, minArea);
            return maxValue;
        }
        return 0;
    }


    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringBuilder sb = new StringBuilder();
        while (true) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            N = Integer.parseInt(st.nextToken());
            if (N == 0) {
                break;
            }
            arr = new int[N];
            for (int i = 0; i < N; i++) {
                arr[i] = Integer.parseInt(st.nextToken());
            }
            tree = new int[4 * N];
            init(1, 0, N - 1);
            long result = getMaxSquare(0, N - 1);
            sb.append(result).append("\n");
        }//end while

        System.out.println(sb);
    }
}
